ext/ffi/lib.rs

```diff
@@ -535,322 +569,809 @@ where
+    .map(|(index, ffi_arg)| unsafe {
+      ffi_arg.as_arg(*parameter_types.get(index).unwrap())
+    })
     .collect();
 
-  let mut native_values: Vec<NativeValue> = vec![];
+  Ok(match result_type {
+    NativeType::Void => NativeValue {
+      void_value: unsafe { cif.call::<()>(fun_ptr, &call_args) },
+    },
+    NativeType::U8 => NativeValue {
+      u8_value: unsafe { cif.call::<u8>(fun_ptr, &call_args) },
+    },
+    NativeType::I8 => NativeValue {
+      i8_value: unsafe { cif.call::<i8>(fun_ptr, &call_args) },
+    },
+    NativeType::U16 => NativeValue {
+      u16_value: unsafe { cif.call::<u16>(fun_ptr, &call_args) },
+    },
+    NativeType::I16 => NativeValue {
+      i16_value: unsafe { cif.call::<i16>(fun_ptr, &call_args) },
+    },
+    NativeType::U32 => NativeValue {
+      u32_value: unsafe { cif.call::<u32>(fun_ptr, &call_args) },
+    },
+    NativeType::I32 => NativeValue {
+      i32_value: unsafe { cif.call::<i32>(fun_ptr, &call_args) },
+    },
+    NativeType::U64 => NativeValue {
+      u64_value: unsafe { cif.call::<u64>(fun_ptr, &call_args) },
+    },
+    NativeType::I64 => NativeValue {
+      i64_value: unsafe { cif.call::<i64>(fun_ptr, &call_args) },
+    },
+    NativeType::USize => NativeValue {
+      usize_value: unsafe { cif.call::<usize>(fun_ptr, &call_args) },
+    },
+    NativeType::ISize => NativeValue {
+      isize_value: unsafe { cif.call::<isize>(fun_ptr, &call_args) },
+    },
+    NativeType::F32 => NativeValue {
+      f32_value: unsafe { cif.call::<f32>(fun_ptr, &call_args) },
+    },
+    NativeType::F64 => NativeValue {
+      f64_value: unsafe { cif.call::<f64>(fun_ptr, &call_args) },
+    },
+    NativeType::Pointer | NativeType::Function {} => NativeValue {
+      pointer: unsafe { cif.call::<*const u8>(fun_ptr, &call_args) },
+    },
+  })
+}
 
-  for (&native_type, value) in symbol
-    .parameter_types
-    .iter()
-    .zip(args.parameters.into_iter())
-  {
-    match native_type {
-      NativeType::Pointer => match value.as_u64() {
-        Some(idx) => {
-          let buf = buffers
-            .get(idx as usize)
-            .ok_or_else(|| {
-              generic_error(format!("No buffer present at index {}", idx))
-            })?
-            .unwrap();
-          native_values.push(NativeValue::buffer(buf.as_ptr()));
-        }
-        _ => {
-          let value = NativeValue::new(native_type, value)?;
-          native_values.push(value);
-        }
-      },
+struct UnsafeCallbackResource {
+  closure: libffi::middle::Closure<'static>,
+  info: *const CallbackInfo,
+}
+
+impl Resource for UnsafeCallbackResource {
+  fn name(&self) -> Cow<str> {
+    "registeredcallback".into()
+  }
+
+  fn close(self: Rc<Self>) {
+    let info = unsafe { Box::from_raw(self.info as *mut CallbackInfo) };
+    let isolate = unsafe { info.isolate.as_mut().unwrap() };
+    unsafe { v8::Global::from_raw(isolate, info.callback) };
+    unsafe { v8::Global::from_raw(isolate, info.context) };
```

Comment:
`close()` is essentially an unsafe operation - It's UB to drop the Global and CallbackInfo while `deno_ffi_callback` references it. Either rename `close` in JS to `unsafeClose` or indicate `deno_ffi_callback` about close() using a bool field in CallbackInfo?

Reply:
Ah yes, in that case `close()` is fine. An invalid pointer is a predictable consequence - let's just document this anyways.